serde: fix c++23 enum handling

dotnwat opened the pull request:
See commit messages. Changes `cluster::incremental_update_operation` to use scoped enum.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat committed:
cluster: make incremental_update_operation scoped enum

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
serde: fix enum handling with c++23

The is_serde check enabled in c++23 (prior to this commit) checked only
for scoped enums. So the SFINAE magic for enum wasn't being triggered
for non-scoped enum. Instead, it was probably being handled by
scalar/integer encoding. Indeed, a not-enum guard is present on scalar
customization points, but, not-enum in c++23, again, means not scoped
enum.

Long ago we informally banned non-scoped enums. Turns out, there was a
non-scoped enum being used in serde. See previous commit for that.

This commit changes serde so that it retains the previous behavior that
worked in c++20, namely, the use of std::is_enum which checks the
superset of scoped and non-scoped enums.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat commented:
@IoannisRP we still haven't looked into the 64-bit-enum-falls-back-to-scalar thing. but that'd be a good one to check out too.

dotnwat commented on src/v/serde/rw/enum.h:
```diff
@@ -24,6 +24,7 @@ namespace serde {
 template<typename T>
 requires(serde_is_enum_v<std::decay_t<T>>)
 void tag_invoke(tag_t<write_tag>, iobuf& out, T t) {
+    static_assert(std::is_scoped_enum_v<std::decay_t<T>>);
```
don't let your friends use non-scoped enums

mmaslankaprv replied:
`template< class T >struct is_scoped_enum;    (since C++23)` will not work with CMake :-(

rockwotj replied:
it works with Bazel, and when I tried to switch cmake to c++23 it broke the upgrade tests because this was an incompatible change to the serialization format for this enum

rockwotj replied:
oh I see, we need to guard the static assert with c++ version checks

bashtanov replied:
Maybe write in a more verbose way that would work with both c++20 and c++23 until we fully switch? [Something like this](https://en.cppreference.com/w/cpp/types/is_scoped_enum#Possible_implementation).

dotnwat replied:
i don't think we need to do anything here, right? we can run CI on the vtools side with c++23 turned on?

rockwotj replied:
Yeah but that's going to require the annoying commit vtools PR and this PR "together" song and dance. I'm fine with doing that (glad that we don't have to do this anymore with Bazel).

dotnwat replied:
LOL i independently used the phrasing "song and dance" in the PR i just pushed which made the same observation

dotnwat replied:
> Maybe write in a more verbose way that would work with both c++20 and c++23 until we fully switch? [Something like this](https://en.cppreference.com/w/cpp/types/is_scoped_enum#Possible_implementation).

Good idea. I think we can compile it out of the tree for now, though, because in principle the fix only needs to last a day or two until vtools switches over to c++23.

bashtanov commented on src/v/cluster/types.h:
```diff
@@ -498,6 +498,11 @@ incremental_update_operation_as_string(incremental_update_operation op) {
+inline std::ostream&
+operator<<(std::ostream& os, const incremental_update_operation& op) {
+    return os << incremental_update_operation_as_string(op);
+}
```
nit: Once we have this operator, can we omit `incremental_update_operation_as_string` function call in `operator<<(std::ostream& o, const property_update<T>& p)` and `operator<<(std::ostream& o, const property_update<tristate<T>>& p)` to get rid of this function?

dotnwat committed:
serde: remove c++23 in c++20

This is so that we can run CI without having to do the song and dance on
vtools side to select the proper dev branch.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>
